In [1]:
#To predict the value of customers: either high value or not

In [2]:
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
import datetime as dt
%matplotlib inline

In [3]:
customer_classes=pd.read_csv(r"E:\My Python codes\Customer_classes.csv", index_col='customer_id')
customer_classes

gender   age state  past_3_years_bike_related_purchases  \
customer_id                                                            
1            Female  64.0   NSW                                   93   
2              Male  37.0   NSW                                   81   
4              Male  56.0   QLD                                   33   
5            Female  40.0   NSW                                   56   
6              Male  51.0   VIC                                   35   
...             ...   ...   ...                                  ...   
3496           Male  31.0   NSW                                   99   
3497         Female  31.0   VIC                                   73   
3498         Female  22.0   VIC                                   28   
3499           Male  38.0   QLD                                   29   
3500         Female  50.0   NSW                                   71   

            job_industry_category     wealth_segment owns_car  tenure  \
customer_id                                                             
1                          Health      Mass Customer      Yes    11.0   
2              Financial Services      Mass Customer      Yes    16.0   
4                              IT      Mass Customer       No     7.0   
5                          Health  Affluent Customer      Yes     8.0   
6                          Retail     High Net Worth      Yes    13.0   
...                           ...                ...      ...     ...   
3496                Manufacturing      Mass Customer      Yes    19.0   
3497                Manufacturing  Affluent Customer      Yes    18.0   
3498                Manufacturing      Mass Customer       No     5.0   
3499                Manufacturing      Mass Customer      Yes     7.0   
3500                Entertainment  Affluent Customer       No    17.0   

             property_valuation       class  
customer_id                                  
1                            10  High Value  
2                            10   Mid Value  
4                             9   Mid Value  
5                             4   Mid Value  
6                             9   Mid Value  
...                         ...         ...  
3496                          9   Mid Value  
3497                          5   Mid Value  
3498                          4   Mid Value  
3499                          9  High Value  
3500                         10   Mid Value  

[3489 rows x 10 columns]

In [4]:
customer_classes.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3489 entries, 1 to 3500
Data columns (total 10 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   gender                               3489 non-null   object 
 1   age                                  3489 non-null   float64
 2   state                                3489 non-null   object 
 3   past_3_years_bike_related_purchases  3489 non-null   int64  
 4   job_industry_category                3489 non-null   object 
 5   wealth_segment                       3489 non-null   object 
 6   owns_car                             3489 non-null   object 
 7   tenure                               3489 non-null   float64
 8   property_valuation                   3489 non-null   int64  
 9   class                                3489 non-null   object 
dtypes: float64(2), int64(2), object(6)
memory usage: 299.8+ KB


In [5]:
customer_classes['gender'].value_counts()

Female    1758
Male      1654
U           77
Name: gender, dtype: int64

In [6]:
jic = pd.get_dummies(customer_classes['job_industry_category'])
jic

Argiculture  Entertainment  Financial Services  Health  IT  \
customer_id                                                               
1                      0              0                   0       1   0   
2                      0              0                   1       0   0   
4                      0              0                   0       0   1   
5                      0              0                   0       1   0   
6                      0              0                   0       0   0   
...                  ...            ...                 ...     ...  ..   
3496                   0              0                   0       0   0   
3497                   0              0                   0       0   0   
3498                   0              0                   0       0   0   
3499                   0              0                   0       0   0   
3500                   0              1                   0       0   0   

             Manufacturing  Property  Retail  Telecommunications  
customer_id                                                       
1                        0         0       0                   0  
2                        0         0       0                   0  
4                        0         0       0                   0  
5                        0         0       0                   0  
6                        0         0       1                   0  
...                    ...       ...     ...                 ...  
3496                     1         0       0                   0  
3497                     1         0       0                   0  
3498                     1         0       0                   0  
3499                     1         0       0                   0  
3500                     0         0       0                   0  

[3489 rows x 9 columns]

In [7]:
customers_dummies = pd.get_dummies(customer_classes, columns=['gender','state','job_industry_category','wealth_segment'])
customers_dummies['owns_car'] = customers_dummies['owns_car'].replace(['Yes','No'],['1','0']).astype(float)
customers_dummies

age  past_3_years_bike_related_purchases  owns_car  tenure  \
customer_id                                                                
1            64.0                                   93       1.0    11.0   
2            37.0                                   81       1.0    16.0   
4            56.0                                   33       0.0     7.0   
5            40.0                                   56       1.0     8.0   
6            51.0                                   35       1.0    13.0   
...           ...                                  ...       ...     ...   
3496         31.0                                   99       1.0    19.0   
3497         31.0                                   73       1.0    18.0   
3498         22.0                                   28       0.0     5.0   
3499         38.0                                   29       1.0     7.0   
3500         50.0                                   71       0.0    17.0   

             property_valuation       class  gender_Female  gender_Male  \
customer_id                                                               
1                            10  High Value              1            0   
2                            10   Mid Value              0            1   
4                             9   Mid Value              0            1   
5                             4   Mid Value              1            0   
6                             9   Mid Value              0            1   
...                         ...         ...            ...          ...   
3496                          9   Mid Value              0            1   
3497                          5   Mid Value              1            0   
3498                          4   Mid Value              1            0   
3499                          9  High Value              0            1   
3500                         10   Mid Value              1            0   

             gender_U  state_NSW  ...  \
customer_id                       ...   
1                   0          1  ...   
2                   0          1  ...   
4                   0          0  ...   
5                   0          1  ...   
6                   0          0  ...   
...               ...        ...  ...   
3496                0          1  ...   
3497                0          0  ...   
3498                0          0  ...   
3499                0          0  ...   
3500                0          1  ...   

             job_industry_category_Financial Services  \
customer_id                                             
1                                                   0   
2                                                   1   
4                                                   0   
5                                                   0   
6                                                   0   
...                                               ...   
3496                                                0   
3497                                                0   
3498                                                0   
3499                                                0   
3500                                                0   

             job_industry_category_Health  job_industry_category_IT  \
customer_id                                                           
1                                       1                         0   
2                                       0                         0   
4                                       0                         1   
5                                       1                         0   
6                                       0                         0   
...                                   ...                       ...   
3496                                    0                         0   
3497                                    0                         0   
3498                                    0                      

In [8]:
#categorise High Value, mid value and low value customers into 2 categories. 
#High value and mid value will be category 1 and low value 0
customers_dummies['class']= customers_dummies['class'].map({'High Value':1, 'Mid Value':1, 'Low Value':0})

In [9]:
customers_dummies['class'].unique()

array([1, 0], dtype=int64)

In [10]:
cols = list(customers_dummies.columns.values)
cols

['age',
 'past_3_years_bike_related_purchases',
 'owns_car',
 'tenure',
 'property_valuation',
 'class',
 'gender_Female',
 'gender_Male',
 'gender_U',
 'state_NSW',
 'state_QLD',
 'state_VIC',
 'job_industry_category_Argiculture',
 'job_industry_category_Entertainment',
 'job_industry_category_Financial Services',
 'job_industry_category_Health',
 'job_industry_category_IT',
 'job_industry_category_Manufacturing',
 'job_industry_category_Property',
 'job_industry_category_Retail',
 'job_industry_category_Telecommunications',
 'wealth_segment_Affluent Customer',
 'wealth_segment_High Net Worth',
 'wealth_segment_Mass Customer']

In [11]:
customers_df =customers_dummies[['age',
 'past_3_years_bike_related_purchases',
 'owns_car',
 'tenure',
 'property_valuation',
 'gender_Female',
 'gender_Male',
 'gender_U',
 'state_NSW',
 'state_QLD',
 'job_industry_category_Financial Services',
 'job_industry_category_Health',
 'job_industry_category_IT',
 'job_industry_category_Property',
 'job_industry_category_Retail',
 'job_industry_category_Telecommunications',
 'wealth_segment_Affluent Customer',
 'wealth_segment_High Net Worth',
 'wealth_segment_Mass Customer','class']]
customers_df

age  past_3_years_bike_related_purchases  owns_car  tenure  \
customer_id                                                                
1            64.0                                   93       1.0    11.0   
2            37.0                                   81       1.0    16.0   
4            56.0                                   33       0.0     7.0   
5            40.0                                   56       1.0     8.0   
6            51.0                                   35       1.0    13.0   
...           ...                                  ...       ...     ...   
3496         31.0                                   99       1.0    19.0   
3497         31.0                                   73       1.0    18.0   
3498         22.0                                   28       0.0     5.0   
3499         38.0                                   29       1.0     7.0   
3500         50.0                                   71       0.0    17.0   

             property_valuation  gender_Female  gender_Male  gender_U  \
customer_id                                                             
1                            10              1            0         0   
2                            10              0            1         0   
4                             9              0            1         0   
5                             4              1            0         0   
6                             9              0            1         0   
...                         ...            ...          ...       ...   
3496                          9              0            1         0   
3497                          5              1            0         0   
3498                          4              1            0         0   
3499                          9              0            1         0   
3500                         10              1            0         0   

             state_NSW  state_QLD  job_industry_category_Financial Services  \
customer_id                                                                   
1                    1          0                                         0   
2                    1          0                                         1   
4                    0          1                                         0   
5                    1          0                                         0   
6                    0          0                                         0   
...                ...        ...                                       ...   
3496                 1          0                                         0   
3497                 0          0                                         0   
3498                 0          0                                         0   
3499                 0          1                                         0   
3500                 1          0                                         0   

             job_industry_category_Health  job_industry_category_IT  \
customer_id                                                           
1                                       1                         0   
2                                       0                         0   
4                                       0                         1   
5                                       1                         0   
6                                       0                         0   
...                                   ...                       ...   
3496                                    0                         0   
3497                                    0                         0   
3498                                    0                         0   
3499                                    0                         0   
3500                                    0                         0   

             job_industry_category_Property  job_industry_category_Retail  \
customer_id                                   

In [ ]:
#Select Target

In [12]:
targets = customers_df['class']
targets

customer_id
1       1
2       1
4       1
5       1
6       1
       ..
3496    1
3497    1
3498    1
3499    1
3500    1
Name: class, Length: 3489, dtype: int64

In [13]:
customers_df['class'].sum()/customers_df['class'].shape[0]

0.9031241043278877

In [14]:
#Select inputs

In [15]:
customers_df.iloc[:,0:20]

age  past_3_years_bike_related_purchases  owns_car  tenure  \
customer_id                                                                
1            64.0                                   93       1.0    11.0   
2            37.0                                   81       1.0    16.0   
4            56.0                                   33       0.0     7.0   
5            40.0                                   56       1.0     8.0   
6            51.0                                   35       1.0    13.0   
...           ...                                  ...       ...     ...   
3496         31.0                                   99       1.0    19.0   
3497         31.0                                   73       1.0    18.0   
3498         22.0                                   28       0.0     5.0   
3499         38.0                                   29       1.0     7.0   
3500         50.0                                   71       0.0    17.0   

             property_valuation  gender_Female  gender_Male  gender_U  \
customer_id                                                             
1                            10              1            0         0   
2                            10              0            1         0   
4                             9              0            1         0   
5                             4              1            0         0   
6                             9              0            1         0   
...                         ...            ...          ...       ...   
3496                          9              0            1         0   
3497                          5              1            0         0   
3498                          4              1            0         0   
3499                          9              0            1         0   
3500                         10              1            0         0   

             state_NSW  state_QLD  job_industry_category_Financial Services  \
customer_id                                                                   
1                    1          0                                         0   
2                    1          0                                         1   
4                    0          1                                         0   
5                    1          0                                         0   
6                    0          0                                         0   
...                ...        ...                                       ...   
3496                 1          0                                         0   
3497                 0          0                                         0   
3498                 0          0                                         0   
3499                 0          1                                         0   
3500                 1          0                                         0   

             job_industry_category_Health  job_industry_category_IT  \
customer_id                                                           
1                                       1                         0   
2                                       0                         0   
4                                       0                         1   
5                                       1                         0   
6                                       0                         0   
...                                   ...                       ...   
3496                                    0                         0   
3497                                    0                         0   
3498                                    0                         0   
3499                                    0                         0   
3500                                    0                         0   

             job_industry_category_Property  job_industry_category_Retail  \
customer_id                                   

In [16]:
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split




In [17]:
unscaled_inputs = customers_df.iloc[:,:19]
unscaled_inputs

age  past_3_years_bike_related_purchases  owns_car  tenure  \
customer_id                                                                
1            64.0                                   93       1.0    11.0   
2            37.0                                   81       1.0    16.0   
4            56.0                                   33       0.0     7.0   
5            40.0                                   56       1.0     8.0   
6            51.0                                   35       1.0    13.0   
...           ...                                  ...       ...     ...   
3496         31.0                                   99       1.0    19.0   
3497         31.0                                   73       1.0    18.0   
3498         22.0                                   28       0.0     5.0   
3499         38.0                                   29       1.0     7.0   
3500         50.0                                   71       0.0    17.0   

             property_valuation  gender_Female  gender_Male  gender_U  \
customer_id                                                             
1                            10              1            0         0   
2                            10              0            1         0   
4                             9              0            1         0   
5                             4              1            0         0   
6                             9              0            1         0   
...                         ...            ...          ...       ...   
3496                          9              0            1         0   
3497                          5              1            0         0   
3498                          4              1            0         0   
3499                          9              0            1         0   
3500                         10              1            0         0   

             state_NSW  state_QLD  job_industry_category_Financial Services  \
customer_id                                                                   
1                    1          0                                         0   
2                    1          0                                         1   
4                    0          1                                         0   
5                    1          0                                         0   
6                    0          0                                         0   
...                ...        ...                                       ...   
3496                 1          0                                         0   
3497                 0          0                                         0   
3498                 0          0                                         0   
3499                 0          1                                         0   
3500                 1          0                                         0   

             job_industry_category_Health  job_industry_category_IT  \
customer_id                                                           
1                                       1                         0   
2                                       0                         0   
4                                       0                         1   
5                                       1                         0   
6                                       0                         0   
...                                   ...                       ...   
3496                                    0                         0   
3497                                    0                         0   
3498                                    0                         0   
3499                                    0                         0   
3500                                    0                         0   

             job_industry_category_Property  job_industry_category_Retail  \
customer_id                                   

In [18]:
X = unscaled_inputs
class_scaler = StandardScaler()
X = class_scaler.fit_transform(X)
y = targets


In [19]:
#Split data into train and test

In [20]:
train_test_split(X, y)

[array([[ 0.88084207,  0.63643344,  0.98718443, ...,  1.76064832,
         -0.58694861, -0.99971343],
        [-1.33394761,  1.16084922, -1.01298194, ..., -0.5679726 ,
         -0.58694861,  1.00028666],
        [-0.93844945,  1.23077133,  0.98718443, ...,  1.76064832,
         -0.58694861, -0.99971343],
        ...,
        [ 0.01074612,  1.44053764,  0.98718443, ..., -0.5679726 ,
         -0.58694861,  1.00028666],
        [-1.17574834,  0.25186186,  0.98718443, ..., -0.5679726 ,
          1.70372667, -0.99971343],
        [ 0.24804502, -1.49619074,  0.98718443, ..., -0.5679726 ,
         -0.58694861,  1.00028666]]),
 array([[-0.70115056, -0.97177496,  0.98718443, ..., -0.5679726 ,
          1.70372667, -0.99971343],
        [-0.06835351,  0.77627765,  0.98718443, ..., -0.5679726 ,
          1.70372667, -0.99971343],
        [-0.62205093, -0.41239813, -1.01298194, ...,  1.76064832,
         -0.58694861, -0.99971343],
        ...,
        [-1.65034613,  1.6153429 , -1.01298194, ..., -

In [21]:
x_train, x_test, y_train, y_test = train_test_split(X, y, train_size=0.80, shuffle=True, random_state=20)

In [22]:
print(x_train.shape, y_train.shape)

(2791, 19) (2791,)


In [23]:
#training inputs are 2791 by 23 while the training targest are 2791(class column)

In [24]:
print(x_test.shape, y_test.shape)

(698, 19) (698,)


In [25]:
#Logistic regression Model

In [26]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

In [27]:
#Training the model

In [28]:
reg = LogisticRegression()

In [29]:
reg.fit(x_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [30]:
reg.score(x_train, y_train)

0.8989609458975277

In [31]:
#Find intercept and coefficients

In [32]:
reg.intercept_

array([2.23303254])

In [33]:
reg.coef_

array([[-0.04385237, -0.02521041, -0.1774947 , -0.01998116,  0.10001503,
        -0.083568  ,  0.10440545, -0.07044955, -0.06972947,  0.01788895,
        -0.01517657, -0.06614175, -0.02033201, -0.09151065, -0.00349423,
         0.06748809,  0.0525247 , -0.08511145,  0.02919895]])

In [34]:
unscaled_inputs.columns.values

array(['age', 'past_3_years_bike_related_purchases', 'owns_car', 'tenure',
       'property_valuation', 'gender_Female', 'gender_Male', 'gender_U',
       'state_NSW', 'state_QLD',
       'job_industry_category_Financial Services',
       'job_industry_category_Health', 'job_industry_category_IT',
       'job_industry_category_Property', 'job_industry_category_Retail',
       'job_industry_category_Telecommunications',
       'wealth_segment_Affluent Customer',
       'wealth_segment_High Net Worth', 'wealth_segment_Mass Customer'],
      dtype=object)

In [35]:
feature_name = unscaled_inputs.columns.values


summary_table = pd.DataFrame(columns=['Feature Name'], data =feature_name)

summary_table['Coefficient']=np.transpose(reg.coef_)# transpose because ndarrays are roes by default
summary_table

summary_table.index= summary_table.index+1
summary_table.loc[0]=['Intercept', reg.intercept_[0]]
summary_table=summary_table.sort_index()
summary_table.sort_values(by=['Coefficient'],ascending=True)

summary_table['Odds Ratio']=np.exp(summary_table.Coefficient)
.sort_values('Odds Ratio', ascending=False)

A feature is not  particularly important if: coefficient is closer to 0 and odds ratio is closer to 1

In [40]:
#Interpretating coefficients

The closer the coefficient is to 0, the smaller the weight
The further away from 0, the bigger the weight
Holds true for standardized models only (when all variables in the same scale)

In [42]:
#Testing the Model

In [43]:
reg.score(x_test, y_test)

0.9197707736389685

In [44]:
predict_proba = reg.predict_proba(x_test)
predict_proba

array([[0.09866499, 0.90133501],
       [0.14139951, 0.85860049],
       [0.10936973, 0.89063027],
       ...,
       [0.0747375 , 0.9252625 ],
       [0.0647623 , 0.9352377 ],
       [0.16559188, 0.83440812]])

If probablility <0.5 = 0
If probability >0.5 = 1

In [46]:
#Saving the model

In [47]:
import pickle

In [48]:
with open('model1', 'wb') as file:
    pickle.dump(reg, file)

In [49]:
with open('scaler1', 'wb') as file:
    pickle.dump(class_scaler, file)

In [50]:
#PREDICT WHICH NEW CUSTOMERS ARE POTENTIAL HIGH VALUE CUSTOMERS

In [51]:
df = pd.read_csv(r"E:\My Python codes\new_customers.csv")
df

name  gender  age  past_3_years_bike_related_purchases  \
0        Chickie Brister    Male   60                                   86   
1           Morly Genery    Male   47                                   69   
2      Ardelis Forrester  Female   43                                   10   
3           Lucine Stutt  Female   38                                   64   
4         Melinda Hadlee  Female   52                                   34   
..                   ...     ...  ...                                  ...   
995  Ferdinand Romanetti    Male   58                                   60   
996         Burk Wortley    Male   16                                   22   
997       Melloney Temby  Female   63                                   17   
998       Dickie Cubbini    Male   65                                   30   
999        Sylas Duffill    Male   62                                   56   

    job_industry_category     wealth_segment owns_car  tenure state  \
0           Manufacturing      Mass Customer      Yes      14   QLD   
1                Property      Mass Customer       No      16   NSW   
2      Financial Services  Affluent Customer       No      10   VIC   
3           Manufacturing  Affluent Customer      Yes       5   QLD   
4      Financial Services  Affluent Customer       No      19   NSW   
..                    ...                ...      ...     ...   ...   
995    Financial Services  Affluent Customer       No       9   NSW   
996                Health      Mass Customer       No       6   NSW   
997    Financial Services  Affluent Customer      Yes      15   QLD   
998    Financial Services      Mass Customer      Yes      19   QLD   
999              Property      Mass Customer      Yes      14   NSW   

     property_valuation  
0                     6  
1                    11  
2                     5  
3                     1  
4                     9  
..                  ...  
995                   7  
996                  10  
997                   2  
998                   2  
999                   9  

[1000 rows x 10 columns]

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
 #   Column                               Non-Null Count  Dtype 
---  ------                               --------------  ----- 
 0   name                                 1000 non-null   object
 1   gender                               1000 non-null   object
 2   age                                  1000 non-null   int64 
 3   past_3_years_bike_related_purchases  1000 non-null   int64 
 4   job_industry_category                1000 non-null   object
 5   wealth_segment                       1000 non-null   object
 6   owns_car                             1000 non-null   object
 7   tenure                               1000 non-null   int64 
 8   state                                1000 non-null   object
 9   property_valuation                   1000 non-null   int64 
dtypes: int64(4), object(6)
memory usage: 78.2+ KB


In [53]:
#Get dummies
df_dummies = pd.get_dummies(df, columns=['gender','state','job_industry_category','wealth_segment']).drop(columns=['job_industry_category_Argiculture','job_industry_category_Manufacturing','state_VIC','job_industry_category_Entertainment'], axis=1)
df_dummies['owns_car'] = df_dummies['owns_car'].replace(['Yes','No'],['1','0']).astype(float)
df_dummies

name  age  past_3_years_bike_related_purchases  owns_car  \
0        Chickie Brister   60                                   86       1.0   
1           Morly Genery   47                                   69       0.0   
2      Ardelis Forrester   43                                   10       0.0   
3           Lucine Stutt   38                                   64       1.0   
4         Melinda Hadlee   52                                   34       0.0   
..                   ...  ...                                  ...       ...   
995  Ferdinand Romanetti   58                                   60       0.0   
996         Burk Wortley   16                                   22       0.0   
997       Melloney Temby   63                                   17       1.0   
998       Dickie Cubbini   65                                   30       1.0   
999        Sylas Duffill   62                                   56       1.0   

     tenure  property_valuation  gender_Female  gender_Male  gender_U  \
0        14                   6              0            1         0   
1        16                  11              0            1         0   
2        10                   5              1            0         0   
3         5                   1              1            0         0   
4        19                   9              1            0         0   
..      ...                 ...            ...          ...       ...   
995       9                   7              0            1         0   
996       6                  10              0            1         0   
997      15                   2              1            0         0   
998      19                   2              0            1         0   
999      14                   9              0            1         0   

     state_NSW  state_QLD  job_industry_category_Financial Services  \
0            0          1                                         0   
1            1          0                                         0   
2            0          0                                         1   
3            0          1                                         0   
4            1          0                                         1   
..         ...        ...                                       ...   
995          1          0                                         1   
996          1          0                                         0   
997          0          1                                         1   
998          0          1                                         1   
999          1          0                                         0   

     job_industry_category_Health  job_industry_category_IT  \
0                               0                         0   
1                               0                         0   
2                               0                         0   
3                               0                         0   
4                               0                         0   
..                            ...                       ...   
995                             0                         0   
996                             1                         0   
997                             0                         0   
998                             0                         0   
999                             0                         0   

     job_industry_category_Property  job_industry_category_Retail  \
0                                 0                             0   
1                                 1                             0   
2                                 0                             0   
3                                 0                             0   
4                                 0                             0   
..                              ...                           ...   
995                               0                             0   
996                 

In [54]:
#STANDARDIZE DATA

In [55]:
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.preprocessing import StandardScaler
   
class CustomScaler(BaseEstimator,TransformerMixin):
 
    def __init__(self,columns,copy=True,with_mean=True,with_std=True):
        self.scaler = StandardScaler(copy,with_mean,with_std)
        self.columns = columns
        self.mean_ = None
        self.var_ = None
   
    def fit(self, X, y=None):
        self.scaler.fit(X[self.columns], y)
        self.mean_ = np.mean(X[self.columns])
        self.var_ = np.var(X[self.columns])
        return self
    
    def transform(self, X, y=None, copy=None):
        init_col_order = X.columns
        X_scaled = pd.DataFrame(self.scaler.transform(X[self.columns]), columns=self.columns)
        X_not_scaled = X.loc[:,~X.columns.isin(self.columns)]
        return pd.concat([X_not_scaled, X_scaled], axis=1)[init_col_order]


In [56]:
unscaled_inputs1 = df_dummies.iloc[:,1:20]
unscaled_inputs1

age  past_3_years_bike_related_purchases  owns_car  tenure  \
0     60                                   86       1.0      14   
1     47                                   69       0.0      16   
2     43                                   10       0.0      10   
3     38                                   64       1.0       5   
4     52                                   34       0.0      19   
..   ...                                  ...       ...     ...   
995   58                                   60       0.0       9   
996   16                                   22       0.0       6   
997   63                                   17       1.0      15   
998   65                                   30       1.0      19   
999   62                                   56       1.0      14   

     property_valuation  gender_Female  gender_Male  gender_U  state_NSW  \
0                     6              0            1         0          0   
1                    11              0            1         0          1   
2                     5              1            0         0          0   
3                     1              1            0         0          0   
4                     9              1            0         0          1   
..                  ...            ...          ...       ...        ...   
995                   7              0            1         0          1   
996                  10              0            1         0          1   
997                   2              1            0         0          0   
998                   2              0            1         0          0   
999                   9              0            1         0          1   

     state_QLD  job_industry_category_Financial Services  \
0            1                                         0   
1            0                                         0   
2            0                                         1   
3            1                                         0   
4            0                                         1   
..         ...                                       ...   
995          0                                         1   
996          0                                         0   
997          1                                         1   
998          1                                         1   
999          0                                         0   

     job_industry_category_Health  job_industry_category_IT  \
0                               0                         0   
1                               0                         0   
2                               0                         0   
3                               0                         0   
4                               0                         0   
..                            ...                       ...   
995                             0                         0   
996                             1                         0   
997                             0                         0   
998                             0                         0   
999                             0                         0   

     job_industry_category_Property  job_industry_category_Retail  \
0                                 0                             0   
1                                 1                             0   
2                                 0                             0   
3                                 0                             0   
4                                 0                             0   
..                              ...                           ...   
995                               0                             0   
996                               0                             0   
997                               0                             0   
998                               0                             0   
999                               1               

In [57]:
#columns_to_ommit = ['owns_car','gender_Female', 'gender_Male', 'gender_U',
                    #'state_NSW', 'state_QLD',
                    #'job_industry_category_Financial Services',
                    #'job_industry_category_Health', 'job_industry_category_IT',
                    #'job_industry_category_Manufacturing',
                    #'job_industry_category_Property', 'job_industry_category_Retail',
                    #'job_industry_category_Telecommunications',
                    #'wealth_segment_Affluent Customer',
                    #'wealth_segment_High Net Worth', 'wealth_segment_Mass Customer']

In [58]:
#columns_to_scale1 = [x for x in unscaled_inputs1.columns.values if x not in columns_to_ommit]

In [59]:
#class_scaler1 = CustomScaler(columns_to_scale1)

In [60]:
#class_scaler1.fit(unscaled_inputs1)

c:\users\user\appdata\local\programs\python\python38\lib\site-packages\sklearn\base.py:193: FutureWarning: From version 0.24, get_params will raise an AttributeError if a parameter cannot be retrieved as an instance attribute. Previously it would return None.
  warnings.warn('From version 0.24, get_params will raise an '


CustomScaler(columns=['age', 'past_3_years_bike_related_purchases', 'tenure',
                      'property_valuation'],
             copy=None, with_mean=None, with_std=None)

In [61]:
#scaled_inputs1 = class_scaler1.transform(unscaled_inputs1)
#scaled_inputs1

age  past_3_years_bike_related_purchases  owns_car    tenure  \
0    0.818236                             1.301669       1.0  0.518807   
1    0.050161                             0.689780       0.0  0.916056   
2   -0.186170                            -1.433838       0.0 -0.275691   
3   -0.481583                             0.509812       1.0 -1.268813   
4    0.345575                            -0.569993       0.0  1.511930   
..        ...                                  ...       ...       ...   
995  0.700071                             0.365838       0.0 -0.474315   
996 -1.781402                            -1.001915       0.0 -1.070189   
997  0.995484                            -1.181883       1.0  0.717432   
998  1.113650                            -0.713967       1.0  1.511930   
999  0.936402                             0.221864       1.0  0.518807   

     property_valuation  gender_Female  gender_Male  gender_U  state_NSW  \
0             -0.506632              0            1         0          0   
1              1.306654              0            1         0          1   
2             -0.869289              1            0         0          0   
3             -2.319918              1            0         0          0   
4              0.581340              1            0         0          1   
..                  ...            ...          ...       ...        ...   
995           -0.143975              0            1         0          1   
996            0.943997              0            1         0          1   
997           -1.957261              1            0         0          0   
998           -1.957261              0            1         0          0   
999            0.581340              0            1         0          1   

     state_QLD  job_industry_category_Financial Services  \
0            1                                         0   
1            0                                         0   
2            0                                         1   
3            1                                         0   
4            0                                         1   
..         ...                                       ...   
995          0                                         1   
996          0                                         0   
997          1                                         1   
998          1                                         1   
999          0                                         0   

     job_industry_category_Health  job_industry_category_IT  \
0                               0                         0   
1                               0                         0   
2                               0                         0   
3                               0                         0   
4                               0                         0   
..                            ...                       ...   
995                             0                         0   
996                             1                         0   
997                             0                         0   
998                             0                         0   
999                             0                         0   

     job_industry_category_Property  job_industry_category_Retail  \
0                                 0                             0   
1                                 1                             0   
2                                 0                             0   
3                                 0                             0   
4                                 0                             0   
..                              ...                           ...   
995                               0                             0   
996                               0                             0   
997                               0                             0   
998                               0     

In [62]:
print(unscaled_inputs1.shape)

(1000, 19)


In [63]:
predict_proba1 = reg.predict_proba(unscaled_inputs1)
predict_proba1

array([[0.90651892, 0.09348108],
       [0.69311364, 0.30688636],
       [0.4135473 , 0.5864527 ],
       ...,
       [0.77967215, 0.22032785],
       [0.83124462, 0.16875538],
       [0.81492412, 0.18507588]])

In [64]:
predict_proba1[:]

array([[0.90651892, 0.09348108],
       [0.69311364, 0.30688636],
       [0.4135473 , 0.5864527 ],
       ...,
       [0.77967215, 0.22032785],
       [0.83124462, 0.16875538],
       [0.81492412, 0.18507588]])

In [65]:
inputs = unscaled_inputs1.copy()
inputs

age  past_3_years_bike_related_purchases  owns_car  tenure  \
0     60                                   86       1.0      14   
1     47                                   69       0.0      16   
2     43                                   10       0.0      10   
3     38                                   64       1.0       5   
4     52                                   34       0.0      19   
..   ...                                  ...       ...     ...   
995   58                                   60       0.0       9   
996   16                                   22       0.0       6   
997   63                                   17       1.0      15   
998   65                                   30       1.0      19   
999   62                                   56       1.0      14   

     property_valuation  gender_Female  gender_Male  gender_U  state_NSW  \
0                     6              0            1         0          0   
1                    11              0            1         0          1   
2                     5              1            0         0          0   
3                     1              1            0         0          0   
4                     9              1            0         0          1   
..                  ...            ...          ...       ...        ...   
995                   7              0            1         0          1   
996                  10              0            1         0          1   
997                   2              1            0         0          0   
998                   2              0            1         0          0   
999                   9              0            1         0          1   

     state_QLD  job_industry_category_Financial Services  \
0            1                                         0   
1            0                                         0   
2            0                                         1   
3            1                                         0   
4            0                                         1   
..         ...                                       ...   
995          0                                         1   
996          0                                         0   
997          1                                         1   
998          1                                         1   
999          0                                         0   

     job_industry_category_Health  job_industry_category_IT  \
0                               0                         0   
1                               0                         0   
2                               0                         0   
3                               0                         0   
4                               0                         0   
..                            ...                       ...   
995                             0                         0   
996                             1                         0   
997                             0                         0   
998                             0                         0   
999                             0                         0   

     job_industry_category_Property  job_industry_category_Retail  \
0                                 0                             0   
1                                 1                             0   
2                                 0                             0   
3                                 0                             0   
4                                 0                             0   
..                              ...                           ...   
995                               0                             0   
996                               0                             0   
997                               0                             0   
998                               0                             0   
999                               1               

In [66]:

unscaled_inputs1 ['Prediction'] = reg.predict(unscaled_inputs1)
unscaled_inputs1

age  past_3_years_bike_related_purchases  owns_car  tenure  \
0     60                                   86       1.0      14   
1     47                                   69       0.0      16   
2     43                                   10       0.0      10   
3     38                                   64       1.0       5   
4     52                                   34       0.0      19   
..   ...                                  ...       ...     ...   
995   58                                   60       0.0       9   
996   16                                   22       0.0       6   
997   63                                   17       1.0      15   
998   65                                   30       1.0      19   
999   62                                   56       1.0      14   

     property_valuation  gender_Female  gender_Male  gender_U  state_NSW  \
0                     6              0            1         0          0   
1                    11              0            1         0          1   
2                     5              1            0         0          0   
3                     1              1            0         0          0   
4                     9              1            0         0          1   
..                  ...            ...          ...       ...        ...   
995                   7              0            1         0          1   
996                  10              0            1         0          1   
997                   2              1            0         0          0   
998                   2              0            1         0          0   
999                   9              0            1         0          1   

     state_QLD  job_industry_category_Financial Services  \
0            1                                         0   
1            0                                         0   
2            0                                         1   
3            1                                         0   
4            0                                         1   
..         ...                                       ...   
995          0                                         1   
996          0                                         0   
997          1                                         1   
998          1                                         1   
999          0                                         0   

     job_industry_category_Health  job_industry_category_IT  \
0                               0                         0   
1                               0                         0   
2                               0                         0   
3                               0                         0   
4                               0                         0   
..                            ...                       ...   
995                             0                         0   
996                             1                         0   
997                             0                         0   
998                             0                         0   
999                             0                         0   

     job_industry_category_Property  job_industry_category_Retail  \
0                                 0                             0   
1                                 1                             0   
2                                 0                             0   
3                                 0                             0   
4                                 0                             0   
..                              ...                           ...   
995                               0                             0   
996                               0                             0   
997                               0                             0   
998                               0                             0   
999                               1               

In [67]:
unscaled_inputs1['Prediction'].sum()/unscaled_inputs1['Prediction'].shape[0]


0.286

In [68]:
df

name  gender  age  past_3_years_bike_related_purchases  \
0        Chickie Brister    Male   60                                   86   
1           Morly Genery    Male   47                                   69   
2      Ardelis Forrester  Female   43                                   10   
3           Lucine Stutt  Female   38                                   64   
4         Melinda Hadlee  Female   52                                   34   
..                   ...     ...  ...                                  ...   
995  Ferdinand Romanetti    Male   58                                   60   
996         Burk Wortley    Male   16                                   22   
997       Melloney Temby  Female   63                                   17   
998       Dickie Cubbini    Male   65                                   30   
999        Sylas Duffill    Male   62                                   56   

    job_industry_category     wealth_segment owns_car  tenure state  \
0           Manufacturing      Mass Customer      Yes      14   QLD   
1                Property      Mass Customer       No      16   NSW   
2      Financial Services  Affluent Customer       No      10   VIC   
3           Manufacturing  Affluent Customer      Yes       5   QLD   
4      Financial Services  Affluent Customer       No      19   NSW   
..                    ...                ...      ...     ...   ...   
995    Financial Services  Affluent Customer       No       9   NSW   
996                Health      Mass Customer       No       6   NSW   
997    Financial Services  Affluent Customer      Yes      15   QLD   
998    Financial Services      Mass Customer      Yes      19   QLD   
999              Property      Mass Customer      Yes      14   NSW   

     property_valuation  
0                     6  
1                    11  
2                     5  
3                     1  
4                     9  
..                  ...  
995                   7  
996                  10  
997                   2  
998                   2  
999                   9  

[1000 rows x 10 columns]

In [ ]:
new_list = pd.merge(unscaled_inputs1, df, on=['customer_id']).set_index('transaction_id').sort_index()
txn_df

In [71]:
merged = unscaled_inputs1.merge(df, left_index=True, right_index=True, how='inner').drop(columns=['gender','age_y','past_3_years_bike_related_purchases_y','job_industry_category','wealth_segment','owns_car_y','tenure_y','state','property_valuation_y'], axis = 1)
merged

age_x  past_3_years_bike_related_purchases_x  owns_car_x  tenure_x  \
0       60                                     86         1.0        14   
1       47                                     69         0.0        16   
2       43                                     10         0.0        10   
3       38                                     64         1.0         5   
4       52                                     34         0.0        19   
..     ...                                    ...         ...       ...   
995     58                                     60         0.0         9   
996     16                                     22         0.0         6   
997     63                                     17         1.0        15   
998     65                                     30         1.0        19   
999     62                                     56         1.0        14   

     property_valuation_x  gender_Female  gender_Male  gender_U  state_NSW  \
0                       6              0            1         0          0   
1                      11              0            1         0          1   
2                       5              1            0         0          0   
3                       1              1            0         0          0   
4                       9              1            0         0          1   
..                    ...            ...          ...       ...        ...   
995                     7              0            1         0          1   
996                    10              0            1         0          1   
997                     2              1            0         0          0   
998                     2              0            1         0          0   
999                     9              0            1         0          1   

     state_QLD  ...  job_industry_category_Health  job_industry_category_IT  \
0            1  ...                             0                         0   
1            0  ...                             0                         0   
2            0  ...                             0                         0   
3            1  ...                             0                         0   
4            0  ...                             0                         0   
..         ...  ...                           ...                       ...   
995          0  ...                             0                         0   
996          0  ...                             1                         0   
997          1  ...                             0                         0   
998          1  ...                             0                         0   
999          0  ...                             0                         0   

     job_industry_category_Property  job_industry_category_Retail  \
0                                 0                             0   
1                                 1                             0   
2                                 0                             0   
3                                 0                             0   
4                                 0                             0   
..                              ...                           ...   
995                               0                             0   
996                               0                             0   
997                               0                             0   
998                               0                             0   
999                               1                             0   

     job_industry_category_Telecommunications  \
0                                           0   
1                                           0   
2                                           0   
3                                           0   
4                                           0   
..                                        ...   
995                                         0 

In [72]:
merged.columns

Index(['age_x', 'past_3_years_bike_related_purchases_x', 'owns_car_x',
       'tenure_x', 'property_valuation_x', 'gender_Female', 'gender_Male',
       'gender_U', 'state_NSW', 'state_QLD',
       'job_industry_category_Financial Services',
       'job_industry_category_Health', 'job_industry_category_IT',
       'job_industry_category_Property', 'job_industry_category_Retail',
       'job_industry_category_Telecommunications',
       'wealth_segment_Affluent Customer', 'wealth_segment_High Net Worth',
       'wealth_segment_Mass Customer', 'Prediction', 'name'],
      dtype='object')

In [74]:
new_list = merged[['name','age_x', 'past_3_years_bike_related_purchases_x', 'owns_car_x',
       'tenure_x', 'property_valuation_x', 'gender_Female', 'gender_Male',
       'gender_U', 'state_NSW', 'state_QLD',
       'job_industry_category_Financial Services',
       'job_industry_category_Health', 'job_industry_category_IT',
       'job_industry_category_Property', 'job_industry_category_Retail',
       'job_industry_category_Telecommunications',
       'wealth_segment_Affluent Customer', 'wealth_segment_High Net Worth',
       'wealth_segment_Mass Customer', 'Prediction']]
new_list

name  age_x  past_3_years_bike_related_purchases_x  \
0        Chickie Brister     60                                     86   
1           Morly Genery     47                                     69   
2      Ardelis Forrester     43                                     10   
3           Lucine Stutt     38                                     64   
4         Melinda Hadlee     52                                     34   
..                   ...    ...                                    ...   
995  Ferdinand Romanetti     58                                     60   
996         Burk Wortley     16                                     22   
997       Melloney Temby     63                                     17   
998       Dickie Cubbini     65                                     30   
999        Sylas Duffill     62                                     56   

     owns_car_x  tenure_x  property_valuation_x  gender_Female  gender_Male  \
0           1.0        14                     6              0            1   
1           0.0        16                    11              0            1   
2           0.0        10                     5              1            0   
3           1.0         5                     1              1            0   
4           0.0        19                     9              1            0   
..          ...       ...                   ...            ...          ...   
995         0.0         9                     7              0            1   
996         0.0         6                    10              0            1   
997         1.0        15                     2              1            0   
998         1.0        19                     2              0            1   
999         1.0        14                     9              0            1   

     gender_U  state_NSW  ...  job_industry_category_Financial Services  \
0           0          0  ...                                         0   
1           0          1  ...                                         0   
2           0          0  ...                                         1   
3           0          0  ...                                         0   
4           0          1  ...                                         1   
..        ...        ...  ...                                       ...   
995         0          1  ...                                         1   
996         0          1  ...                                         0   
997         0          0  ...                                         1   
998         0          0  ...                                         1   
999         0          1  ...                                         0   

     job_industry_category_Health  job_industry_category_IT  \
0                               0                         0   
1                               0                         0   
2                               0                         0   
3                               0                         0   
4                               0                         0   
..                            ...                       ...   
995                             0                         0   
996                             1                         0   
997                             0                         0   
998                             0                         0   
999                             0                         0   

     job_industry_category_Property  job_industry_category_Retail  \
0                                 0                             0   
1                                 1                             0   
2                                 0                             0   
3                                 0                             0   
4                                 0                             0   
..                              ...                           ...   
995                               0      

In [75]:
new_list['Prediction'].value_counts()

0    714
1    286
Name: Prediction, dtype: int64

In [76]:
#Filter dataframe to get only customers predicted as high value
high_value_customers=new_list[new_list['Prediction']==1]
high_value_customers

name  age_x  past_3_years_bike_related_purchases_x  \
2    Ardelis Forrester     43                                     10   
6       Rutledge Hallt     41                                     23   
11     Wheeler Winward     18                                     48   
12                Olag     27                                     60   
13      Melba Spellacy     41                                     38   
..                 ...    ...                                    ...   
974   Aundrea Outridge     16                                     77   
976      Esme Pilipets     50                                     15   
987       Lolly Prewer     20                                     44   
994     Datha Fishburn     27                                     15   
996       Burk Wortley     16                                     22   

     owns_car_x  tenure_x  property_valuation_x  gender_Female  gender_Male  \
2           0.0        10                     5              1            0   
6           0.0         8                     7              0            1   
11          0.0        10                     8              0            1   
12          0.0         9                    11              0            1   
13          0.0         4                    10              1            0   
..          ...       ...                   ...            ...          ...   
974         1.0        10                    12              1            0   
976         1.0         5                     9              0            1   
987         0.0        15                     7              1            0   
994         0.0         3                    12              1            0   
996         0.0         6                    10              0            1   

     gender_U  state_NSW  ...  job_industry_category_Financial Services  \
2           0          0  ...                                         1   
6           0          1  ...                                         1   
11          0          0  ...                                         0   
12          0          1  ...                                         0   
13          0          1  ...                                         0   
..        ...        ...  ...                                       ...   
974         0          0  ...                                         1   
976         0          0  ...                                         0   
987         0          1  ...                                         0   
994         0          0  ...                                         0   
996         0          1  ...                                         0   

     job_industry_category_Health  job_industry_category_IT  \
2                               0                         0   
6                               0                         0   
11                              0                         0   
12                              0                         0   
13                              1                         0   
..                            ...                       ...   
974                             0                         0   
976                             1                         0   
987                             1                         0   
994                             0                         0   
996                             1                         0   

     job_industry_category_Property  job_industry_category_Retail  \
2                                 0                             0   
6                                 0                             0   
11                                0                             0   
12                                0                             0   
13                                0                             0   
..                              ...                           ...   
974                               0                            

In [78]:
high_value_customers.describe()

age_x  past_3_years_bike_related_purchases_x  owns_car_x  \
count  286.000000                             286.000000  286.000000   
mean    31.402098                              29.349650    0.433566   
std     11.527657                              21.465683    0.496436   
min     15.000000                               0.000000    0.000000   
25%     21.000000                              11.000000    0.000000   
50%     28.500000                              25.000000    0.000000   
75%     41.000000                              45.000000    1.000000   
max     63.000000                              84.000000    1.000000   

         tenure_x  property_valuation_x  gender_Female  gender_Male  \
count  286.000000            286.000000     286.000000   286.000000   
mean     9.220280              8.157343       0.458042     0.527972   
std      4.639692              2.430739       0.499110     0.500092   
min      1.000000              1.000000       0.000000     0.000000   
25%      5.000000              7.000000       0.000000     0.000000   
50%      9.000000              8.000000       0.000000     1.000000   
75%     13.000000             10.000000       1.000000     1.000000   
max     22.000000             12.000000       1.000000     1.000000   

         gender_U   state_NSW   state_QLD  \
count  286.000000  286.000000  286.000000   
mean     0.013986    0.527972    0.174825   
std      0.117638    0.500092    0.380483   
min      0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000   
50%      0.000000    1.000000    0.000000   
75%      0.000000    1.000000    0.000000   
max      1.000000    1.000000    1.000000   

       job_industry_category_Financial Services  job_industry_category_Health  \
count                                286.000000                    286.000000   
mean                                   0.265734                      0.195804   
std                                    0.442498                      0.397514   
min                                    0.000000                      0.000000   
25%                                    0.000000                      0.000000   
50%                                    0.000000                      0.000000   
75%                                    1.000000                      0.000000   
max                                    1.000000                      1.000000   

       job_industry_category_IT  job_industry_category_Property  \
count                286.000000                      286.000000   
mean                   0.045455                        0.052448   
std                    0.208664                        0.223319   
min                    0.000000                        0.000000   
25%                    0.000000                        0.000000   
50%                    0.000000                        0.000000   
75%                    0.000000                        0.000000   
max                    1.000000                        1.000000   

       job_industry_category_Retail  job_industry_category_Telecommunications  \
count                    286.000000                                286.000000   
mean                       0.083916                                  0.034965   
std                        0.277748                                  0.184013   
min                        0.000000                                  0.000000   
25%                        0.000000                                  0.000000   
50%                        0.000000                                  0.000000   
75%                        0.000000                                  0.000000   
max                        1.000000                                  1.000000   

       wealth_segment_Affluent Customer  wealth_segment_High Net Worth  \
count                        286.000000                     286.000000   
mean                           0.265734                       0.230769   
std                            

There are 286 possible high and mid value customers from the list